### DISCUSSION TITLE
Promotions API

### DISCUSSION THREAD

**Discussion by srindom** [medusajs/medusa#1037] [RFC]
Hey everyone! 💜

We are looking into expanding upon the existing promotions capabilities in Medusa and we would love to get some feedback and inputs from you guys. Development on this feature will begin in the near future so the goal is to have a pretty clear idea about what the new API should look like within a week's time.

The current discounts API in Medusa allows you to do simple promotions for a store. Currently, we support:

- percentage discounts (available for multi-regions)
- fixed amount discounts (only 1 region can be selected)
- free shipping discounts (waives the shipping costs on an entire order)
- discounts can be applied on the item level or on the order total (allocation mode)
- in both allocation modes it is possible to define which items the discount works for
- furthermore, we have template (dynamic) codes, scheduling and usage limiting

The above options are pretty standard across most platforms and are also sufficient in many cases; however, we would like to expand upon the existing API to make promotions in Medusa even more powerful. Below are some of the cases that we would like to support going forward:

## Stories

- As a merchant I want to offer discounts for my VIP customers.
- As a merchant I want to have special sale prices in a period of time.
- As a merchant I want to use the most powerful tool for managing promotions (e.g. 3rd party tools)
- As a developer I want to easily integrate multiple different promotions solutions.
- As a developer I want to be able to use a single endpoint for applying promotions instead of having to deal with different APIs.

Enabling the use cases above would make Medusa extremely powerful by default and would also allow 3rd parties to hook into areas where niche use cases are better managed through other tools. 

## Success criteria

The goal for the new promotions API is the have a solution that makes the vast majority of promotions use cases easily attainable; while being remaining flexible enough to accommodate more advanced solutions with a bit extra work.

## What to build

Some of the rough bigger ticket items to implement are as follows:

- customer roles to allow customers to get things like “VIP”
- promotion interface to allow “promotion plugins” should prob contain something like `redeemPromotion` which can validate a given code against a 3rd party system. This function should be called as part of the `addDiscount` flow.
- a way for prices to take over from regular prices - potentially also segmented by customer roles:
    - Could be something like when authenticated customer requests `GET /store/products` the api could format the response such that `product.variants[0].prices` are “promotion prices” and not the regular prices.
- create `SystemPromotionService` which is a default implementation for promotions

## Considerations

- How much should be kept in Medusa vs how much should be controlled externally - i.e. if using a 3rd party tool should we store a “copy” of the code in Medusa or should Medusa only have a notion of externally existing coupons when calling `redeemPromotion`
    - Consider the application of a promotion:
        - `POST /carts/:id/discounts` w. `{ code: "XYZ" }`
        - Should we do:
            - lookup code in Medusa - if it exists apply as normal
            - if it doesn’t exist ask configured promotions plugins if they know about the “XYZ” code and only then allow the code to be applied
        - Or alternatively:
            - create codes in Medusa when created in external systems such that codes exist like so: `{ code: "XYZ", provider_id: "promotion-plugin-provider" }`
- What should the `IPromotionService` interface look like? `redeemPromotion(code, cart, context)` is a good starting point are there other things to consider?
- My thinking is that `redeemPromotion` will respond with a “Medusa-compatible” format. We need to figure out what this format should look like. Could be something like this:
    
    ```markdown
    allocation: "item" | "total"
    type: "percentage" | "fixed"
    value: [what percentage/how much of fixed]
    ```
    
    The goal of this format is primarily to allow the `TotalsService` to compute the totals. Which brings us to the next point.
    
- How should promotions be dealt with in totals computations. A thought that I think could be powerful would be to introduce `line_item_adjustments` these would adjust the regular line item total by some amount. A rough data model could be
    
    ```markdown
    id
    item_id
    type: e.g. "discount" 
    amount: [the amount to adjust the original total by]
    ```
    
    This would allow us to not have to compute the discount all everytime a total is requested.
    
- What rules should be added on top of the existing rule types (r.n. we only have product id)? I think we should add an option for adding promotions based on collection, product tag, product type too.

## Research resources

- [talon.one](http://talon.one) - an api based promotion engine; worth reading into

**Comment by khawarizmus** [comment]
Not sure if it should be part of the promotions API but looking at the resource attached a very nice use case to have is of the referrals. https://www.talon.one/referrals

**Reply by srindom** [reply]
Yess! Very good idea - I think it would be nice to make sure that we can support flows like this; at least initially through 3rd parties, but potentially in the future within the core of Medusa. 

We will need `DiscountRule`s to be more generic and flexible than what they are today. I have been thinking about doing something like this:

```
Discount
- id 
- code
- provider_id
- rules <-- DiscountRules

DiscountRule
- discount_id
- resource: "product" | "collection" | "product tag" ....
- operator: "eq" | "in" | "neq" | "gt" ....
- value
```

But this has a pretty big disadvantage in the sense that you can't query and validate the discounts easily with SQL.

**Reply by mohamed-achich** [reply]
hello @srindom, is it now possible to integrate a 3rd party tools to implement a referral program ?

**Comment by pKorsholm** [comment]
wgrt the challenge you highlight @srindom it seems like there are generally two solutions to storing the discounts with relations: 

**Solution A.**

One way this could be achieved could be to look at the new tax-api for inspiration. In the tax-api two different ways of applying tax exist, it can be applied by product-type or by product directly. 

This is structured in a way where a [tax_rate](https://github.com/medusajs/medusa/blob/8d8d71708fb08d97c0fa388ea517232302b06143/packages/medusa/src/models/tax-rate.ts) is coupled to a product, product_type or shipping option by either a [product-type-tax-rate](https://github.com/medusajs/medusa/blob/tax-api/packages/medusa/src/models/product-type-tax-rate.ts)(https://github.com/medusajs/medusa/blob/tax-api/packages/medusa/src/models/product-type-tax-rate.ts), [product-tax-rate](https://github.com/medusajs/medusa/blob/tax-api/packages/medusa/src/models/product-tax-rate.ts)(https://github.com/medusajs/medusa/blob/tax-api/packages/medusa/src/models/product-tax-rate.ts)  or [shipping-tax-rate](https://github.com/medusajs/medusa/blob/tax-api/packages/medusa/src/models/shipping-tax-rate.ts)(https://github.com/medusajs/medusa/blob/tax-api/packages/medusa/src/models/shipping-tax-rate.ts) entity. 

This concept could be extended with `discount_rule` taking the role of a `tax_rate` and thus will hold many-to-many references to both valid collections, customer groups and other relevant entities. 

In the totals calculation the additional rules would be evaluated in a similar way to how `valid_for` is validated at the moment: 

```jsx
for (const item of cart.items) {
  if (discount.rule.valid_for?.length > 0) {
    discount.rule.valid_for.map(({ id }) => {
      if (item.variant.product_id === id) {
        discounts.push(
          this.calculateDiscount_(
            item,
            item.variant.id,
            item.unit_price,
            discount.rule.value,
            discount.rule.type
          )
        )
      }
    })
  }
}
```

This would additional loops to be added for different discount types as is done specifically with the `valid_for` field above. However some resemblance of this pattern will probably be unavoidable no matter which solution is chosen.

**Suggestion 2**

A different take on a solution would be to create a more generic discount condition element with an enum value describing the condition applied to the discount i.e. product-type, collection, product, customergroup etc.

The object would also contain an identifier of the given element, collection, etc. thus it would look something like this: 

```jsx
DiscountCondition{ 
	type: 'product' | 'collection' | 'customergroup'
	value: string // ids, amounts, etc. 
	condition: 'lt' | 'gt' | 'eq' | etc...
}
```

In this way a discount rule would be extended with an array of conditions which could then be checked when the discount is applied to the cart in the totals service in somewhat the same way as above. 

This solution, though more generic could require additional database queries in order to properly validate the condition if the given information is not present in the cart object when applying the discount. 

**Applying generalised discounts**

When looking into applying discounts we are considering how we can do so in an extendable way to further empower developers to integrate with external tools (such as [[talon.one](http://talon.one/)](http://talon.one)) or to create their own custom validation in the most effortless way.

To achieve this abstraction we could again look to the existing generalised implementations in Medusa core. We have already implemented generalised strategies for cart completion and tax calculation, all of which can be found in the `src/strategies` folder in the core project. 

Using this same pattern we will implement a DiscountCalculationStrategy interface along with a standard BaseDiscountCalculationStrategy which contains some basic logic for discount calculation but which can also be overwritten to implement discount calculation using services such as [[Talon.one](http://talon.one/)](http://Talon.one) as separate plugins.

**Saving state**

Because implementing discount calculations like `valid_for` is not the most efficient computations and computing it every time we get the order total would result in a lot of additional computations, we need a way to store discounted values closer to the lineitems themselves. For this we thought of introducing an `allocation` or `lineItemAdjustment` array on each line item describing each applied adjustment to the line item price i.e. applied discounts, taxes etc. 

Initially the implementation of `lineItemAdjustments` would not replace the need to recompute discounts due to the different constraints imposed by each discount rule which can apply for different scenarios. Thus recomputing present discounts would be required while creating the cart. However once the cart is completed and an order is created the `lineItemAdjustments` can then be created to make totals calculations easier for each order.

**Comment by osseonews** [comment]
One of the most important things for any e-commerce store that most headless platforms miss is: tier pricing. This refers to a discount when you buy a different quantity of a specific SKU, e.g. buy 1 of ABC for $10 each, but buy 5 of ABC for $5 each, buy 10 of ABC for $3 each. I'm not sure if this should be part of promotions or if you add to the pricing schema for a SKU, but it is critical for most businesses. I think you can accomplish something like this with promotions by having a "Buy X, get Y Free" type of promotion.

**Reply by Rask777** [reply]
Were you able to implement it in medusa? different pricing for different quantities

**Reply by Gustavodacrvi** [reply]
Any update on this feature?

**Comment by femisshingala** [comment]
Hello good morning sir

**Reply by pKorsholm** [reply]
Hello, Feel free to reach out if we can help out with any questions regarding medusa 😄

**Comment by popowych** [comment]
Hi there, where are we with those coupon / promotion ideas? Actually it's not possible to
- set up a limit of use per customer 
- authorize the use of a coupon based on figures ... this is probably one of the main reason you want to implement coupon actually. Stories go like "if the cart is over X, then Z", "if that customer buys X items, then he gets Y something"
- set fixed amount for your coupon. What's the commercial logic behind "we'll calculate the coupon amount according to the VAT rate" -- it's a commercial non sense (think of the ADS online). Average consumers see a incl. VAT price, don't care about the excl. VAT one, so why should we based any price / coupon / whatever on it. My commercial discours will use the last price, not the reduced one. 
_the previous point leads to coupon amount fluctuation in your cart based on the number of items, probably a bug_
